Introduce conversion into Viper functions

AlexFalter opened the pull request:
This PR introduces conversion from Kotlin functions annotated as Pure into Viper functions:

For this I introduced two new embeddings:
- PureFunctionEmbedding
- PureFunctionBodyEmbedding
which are similar to their non-pure counter-parts (i.e. FunctionEmbedding and FunctionBodyEmbedding), but embed an expression rather than a statement sequence.
I also introduced the UserFunction in the Viper AST to represent such pure user-defined functions.
Further, I introduced a function map (similar to the methods map) in the program of the ProgramConverter to store embedded functions.

Regarding conversion flow:

The distinction between pure vs. non-pure is maked in the ProgramConverter when embedding the provided function. The function is then embedded in the newly introduced pure embeddings and stored in the function map. Note, that currently the body of the function is being - regardless of input - converted into the integer literal 42.

Testing:

I have introduced a test case ensuring, that @Pure annotated functions are being converted into Viper functions rather than methods. All other test-cases not failing also shows, that non-annotated functions are not being wrongly converted

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/conversion/ProgramConverter.kt:
```diff
@@ -120,9 +125,23 @@ class ProgramConverter(
-            body = stmtCtx.convertMethodWithBody(declaration, signature, returnTarget)
+        if (declaration.symbol.isPure(session)) {
+            embedPureUserFunction(declaration.symbol, signature).apply {
+                body = PureFunctionBodyEmbedding(Exp.IntLit(42), returnTarget)
```
Please leave a TODO here to make it clear that this is a placeholder.  (Preferably, also make a variable that's called `placeholder` so that the extent of the placeholder is clear.)

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/embeddings/callables/FullNamedFunctionSignature.kt:
```diff
@@ -91,4 +94,17 @@ fun FullNamedFunctionSignature.toViperMethod(
+    formalArgs.map { it.toLocalVarDecl() },
+    Type.Int,
+    getPreconditions().pureToViper(toBuiltin = true),
+    emptyList(),
```
Why no postconditions?

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/embeddings/callables/FunctionEmbedding.kt:
```diff
@@ -25,3 +31,16 @@ class UserFunctionEmbedding(private val callable: RichCallableEmbedding) : Funct
+class PureUserFunctionEmbedding(private val callable: RichCallableEmbedding) : PureFunctionEmbedding,
+        CallableEmbedding by callable {
+
+    var body: PureFunctionBodyEmbedding? = null
```
Until you need a PureFunctionBody embedding, perhaps just an `ExpEmbedding` would be best here?

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/embeddings/callables/FunctionEmbedding.kt:
```diff
@@ -25,3 +31,16 @@ class UserFunctionEmbedding(private val callable: RichCallableEmbedding) : Funct
+    var body: PureFunctionBodyEmbedding? = null
+
+    override val viperFunction: Function?
+        get() = body?.toViperFunction(callable) ?: callable.toViperFunctionHeader()
```
I think this is overly permissive - the idea of having a nullable bodyembedding for methods is that we only want to verify some methods.  But at the moment we want to translate all functions to ones with bodies, so that's quite a different story.

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/conversion/ProgramConverter.kt:
```diff
@@ -120,9 +125,23 @@ class ProgramConverter(
         // place the embedding in the map before processing the body.
-        embedUserFunction(declaration.symbol, signature).apply {
-            body = stmtCtx.convertMethodWithBody(declaration, signature, returnTarget)
+        if (declaration.symbol.isPure(session)) {
```
I'm afraid this can't be the right place to do this.  Namely: this is only called for the "root" function that we are verifying, while you want to add bodies to all pure functions (those bodies are visible in methods that call them).  So you want to hook this up in a place that is called for all functions a user calls.

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/embeddings/callables/NonInlineNamedFunction.kt:
```diff
@@ -24,4 +25,10 @@ class NonInlineNamedFunction(val signature: FullNamedFunctionSignature) : RichCa
             PlaceholderVariableEmbedding(PlaceholderReturnVariableName, signature.callableType.returnType)
         )
+
+    override fun toViperFunctionHeader(): Function =
```
I don't think you need this

jesyspa commented on formver.compiler-plugin/testData/diagnostics/conversion/pure_function.kt:
```diff
@@ -0,0 +1,7 @@
+fun <!VIPER_TEXT!>emptyFunction<!>() { }
+
+@Pure
+fun <!VIPER_TEXT!>emptyAnnotatedFunction<!>() { }
```
If you're going to put placeholders, I would put ones where the behaviours are as you expect them to be.  Now, the translation to Viper looks like it is doing something very strange.  (Or at least leave a comment, but given how easy it is to return 42 here, I would do that.)

jesyspa reviewed (changes requested):
This looks like the right direction, some minor comments.  The only big thing is that we need to find a different place to hook this into - give it a shot (add a test where you call the pure function and make sure it is generated), and if you can't find one we can call and look at it together).

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/embeddings/FunctionBodyEmbedding.kt:
```diff
@@ -20,3 +23,12 @@ data class FunctionBodyEmbedding(
         signature.toViperMethod(viperBody, returnTarget.variable)
 }
+
+data class PureFunctionBodyEmbedding(
```
I don't think you need that in a function? The return target is the variable to assign to in order to return a value, but IIRC in functions that is not allowed, the return value is just the result of the body expression.

AlexFalter commented:
Latest push introduced:

- FunctionCall embedding to represent a call to a function
- Added TODO markers
- Changed functions to be translated with return type Ref (see Slack for more details)
- Therefore functions now return null always (isntead of 42)
- Represent function call as an assign statement

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/conversion/StmtConversionVisitor.kt:
```diff
@@ -281,18 +283,40 @@ object StmtConversionVisitor : FirVisitor<ExpEmbedding, StmtConversionContext>()
-                    data,
-                    data.embedType(functionCall.resolvedType),
-                )
+                if (data.isPureFunction(symbol)) {
```
I think this separation of responsibilities is not ideal.  The StmtConversionVisitor should not care whether the function it embeds is pure or not; it just needs to get a callable object from the ProgramConverter, and then put a call in it.  Deciding how to embed that function shouldn't be up to it.

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/embeddings/callables/NonInlineNamedFunction.kt:
```diff
@@ -7,21 +7,34 @@ package org.jetbrains.kotlin.formver.core.embeddings.callables
         ctx: StmtConversionContext,
-    ): ExpEmbedding = MethodCall(signature, args)
+    ): ExpEmbedding {
+        return if (ctx.isPureFunction(signature.symbol)) {
```
I think we should already be aware of this earlier, probably when we create this `NonInlineNamedFunction`, or perhaps even when we create the `signature`

jesyspa commented on formver.compiler-plugin/testData/diagnostics/conversion/pure_function.fir.diag.txt:
```diff
@@ -0,0 +1,18 @@
+/pure_function.kt:(53,66): info: Generated Viper text for emptyFunction:
+function f$emptyAnnotatedFunction$TF$(): Ref
```
Something is going wrong here: you want to include the body.

AlexFalter replied:
This is just the header of the function called in the root function we are translating. If I understand correctly it is done for methods as well (i. e. when called only the header is being generated rather than the whole method again), so I kept this behaviour for now. It looks strange due to missing Pre/Post-Conditions...

jesyspa replied:
That is correct; for methods, we do not generate their bodies, since those are opaque and they are verified separately.  However, this isn't the case for functions: there, the bodies are important since Viper uses them, so a function should include its body.

That said, this can be left as a TODO for now

AlexFalter commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/conversion/StmtConversionVisitor.kt:
```diff
@@ -281,18 +283,40 @@ object StmtConversionVisitor : FirVisitor<ExpEmbedding, StmtConversionContext>()
             }
         }
 
+    private fun List<FirExpression>.withPureVarargsHandled(
```
I thought of this as well, however with the viperMethod being a member of the Function embedding it feels strange to have PureFunctionEmbedding being a subclass of FunctionEmbedding

AlexFalter commented on formver.compiler-plugin/viper/src/org/jetbrains/kotlin/formver/viper/ast/Function.kt:
I agree - however I think I have dug the hole to deep for me to untangle this PR before merge. I have understimated the amount of code changes I need to do for this a bit. So (assuming I get the rest above to work) if alright with you I would keep it in this PR and then be a more considerate of this with the next features

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/conversion/ProgramConverter.kt:
```diff
@@ -178,6 +195,29 @@ class ProgramConverter(
+        }
+    }
+
+    override fun embedAnyFunction(symbol: FirFunctionSymbol<*>): CallableEmbedding {
```
nit: instead of

```kotlin
fun f() {
  return ...
}
```

consider 

```kotlin
fun f() = ...
```

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/embeddings/expression/ControlFlow.kt:
```diff
@@ -222,6 +223,23 @@ data class MethodCall(val method: NamedFunctionSignature, val args: List<ExpEmbe
+
+    override fun <R> accept(v: ExpVisitor<R>): R =
+        v.visitFunctionCall(this) // or however your visitor is set up
+
```
nit: extra newline

jesyspa replied:
Unfortunately no, and it's not always an error - Kotlin doesn't have a strict policy, but it's good to be consistent and to keep the code reasonably compact (as it tends to be quite lengthy already)

jesyspa commented on formver.compiler-plugin/viper/src/org/jetbrains/kotlin/formver/viper/ast/Function.kt:
```diff
@@ -67,6 +67,21 @@ interface Function : IntoSilver<viper.silver.ast.Function>, Applicable {
+) : Function {
+    override val includeInDumpPolicy: IncludeInDumpPolicy = IncludeInDumpPolicy.ALWAYS
+}
+
```
nit: extra newline

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/embeddings/callables/FullNamedFunctionSignature.kt:
```diff
@@ -91,4 +90,17 @@ fun FullNamedFunctionSignature.toViperMethod(
+) = UserFunction(
+    name,
+    formalArgs.map { it.toLocalVarDecl() },
+    // TODO: Be explicit about the return types of functions instead of boxing them into a Ref
```
This is optimistic at this stage; `Ref` is our standard type to pass things around, so this TODO would have to be everywhere if we wanted to mark all the places.

jesyspa commented on formver.compiler-plugin/core/src/org/jetbrains/kotlin/formver/core/embeddings/expression/ControlFlow.kt:
```diff
@@ -222,6 +223,23 @@ data class MethodCall(val method: NamedFunctionSignature, val args: List<ExpEmbe
+    )
+
+    override fun <R> accept(v: ExpVisitor<R>): R =
+        v.visitFunctionCall(this) // or however your visitor is set up
```
I don't follow, what does this comment mean?
